OrbitBase/include/OrbitBase/ThreadPool.h

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) 2020 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef ORBIT_BASE_THREAD_POOL_H_
+#define ORBIT_BASE_THREAD_POOL_H_
+
+#include <memory>
+
+#include "OrbitBase/Action.h"
+
+// This class implements a Thread Pool. Thread Pool allows to execute
+// actions in another thread without need to manage creation and destruction
+// of that thread.
+//
+// Usage example:
+//
+// /* Schedule an action on the thread-pool */
+// thread_poool->Schedule(CreateAction([data]{
+//   DoSomethingWith(data);
+// }));
+//
+// /* Shutdown thread-pool, blocks until all currently executed actions are
+//    complete and stops all worker threads */
+//
+// thread_pool->Shutdown();
+// thread_pool->Wait();
+//
+class ThreadPool {
+ public:
+  ThreadPool() = default;
+  virtual ~ThreadPool() = default;
+
+  // Schedules the action to be performed on the main thread.
+  virtual void Schedule(std::unique_ptr<Action> action) = 0;
+
+  template <typename F>
+  void Schedule(F&& functor) {
+    Schedule(CreateAction(std::forward<F>(functor)));
+  }
+
+  // Initiates shutdown, any Schedule after this call will fail.
+  virtual void Shutdown() = 0;
+
+  // Wait until all tasks are complete. Call this after Shutdown
+  virtual void Wait() = 0;
+
+  // Create ThreadPool with specified minimum and maximum number of worker
+  // threads.
+  //
+  // The queue_threshold is used as a signal to create more worker
+  // threads to manage increased load.
+  //
+  // Whenever an action is Scheduled the thread pool puts it in internal queue.
+  // Worker threads pick actions from the queue and execute them. If at any
+  // point number of actions in the queue reaches queue_threshold, the thread
+  // pool creates new worker thread if current number of worker threads is less
+  // than maximum pool size.
+  //
+  // TODO: If queue is empty thread_pool reduces number of worker threads
```

Comment:
What would trigger the reduction of the size? When a thread is done with a task and the queue is empty, will that thread get destroyed immediately or would it be reasonable to add a wait period?

Reply:
some waiting period - blocking read with timeout -> check if we can reduce number of threads when timed out.